lti_consumer/lti_consumer.py

```diff
@@ -740,6 +799,85 @@ def student_view(self, context):
         fragment.initialize_js('LtiConsumerXBlock')
         return fragment
 
+    has_author_view = True
+
+    def author_view(self, context):
+        """
+        XBlock studio view of this component.
+
+        Makes a request to `lti_launch_handler` either
+        in an iframe or in a new window depending on the
+        configuration of the instance of this XBlock
+
+        Arguments:
+            context (dict): XBlock context
+
+        Returns:
+            xblock.fragment.Fragment: XBlock HTML fragment
+        """
+        if self.lti_version != "lti_1p3":
+            return self.student_view(context)
+
+        fragment = Fragment()
+        loader = ResourceLoader(__name__)
+        context = {
+            "client": self.lti_1p3_client_id,
+            "deployment_id": "1",
+            "keyset_url": get_lms_lti_keyset_link(self.location),  # pylint: disable=no-member
+            "oidc_callback": get_lms_lti_launch_link(),
+            "launch_url": self.lti_1p3_launch_url,
+        }
+        fragment.add_content(loader.render_mako_template('/templates/html/lti_1p3_studio.html', context))
+        fragment.add_css(loader.load_unicode('static/css/student.css'))
+        fragment.add_javascript(loader.load_unicode('static/js/xblock_lti_consumer.js'))
+        fragment.initialize_js('LtiConsumerXBlock')
+        return fragment
+
+    def _get_lti1p3_consumer(self):
+        """
+        Returns LTI 1.3 Consumer class
+
+        Instanced from parameters from XBlock
+        """
+        return LtiConsumer1p3(
```

Comment:
Does it matter that this is creating a new instance of `LtiConsumer1p3` each time `_get_lti1p3_consumer` is called?  Naively, I might have expected it was returning an instance variable.  I don't know that it gets called more than once for a request, but if so, are there any consequences for having separate instances?  LtiConsumer1p3 seems to have instance variables, but I'm not sure if they ever change from what it's instantiated with, so it may (technically) be fine.  Feels slightly odd, though.

Reply:
I've split this function in order to avoid code duplication in the multiple places that it needs to be used. This LTI consumer class is stateless, and doesn't store any request/launch data. It's main function is to receive LTI related variables and generate an output from them.

Do you have an idea of how we can improve this? 
Do you think a property will do better here?

>I don't know that it gets called more than once for a request, but if so, are there any consequences for having separate instances?

We shouldn't see any consequences of this as long as the classes are instantiated in the same way.
I have a small worry about instancing and storing this class since the user might change some parameter in the settings and the saved class might end up with old information (this issue will only happen in the Studio, so it shouldn't cause any trouble, but it's worth mentioning).

Reply:
If it's stateless, I'm with you. 👍

Reply:
@davidjoy As we discussed in the meeting today: 
It isn't exactly a stateless class, since it has a few methods to set user and configuration variables, but it should be used as one on most cases (by instancing a new class, setting variables and running desired LTI method).

I'll add some comments on the code to clarify how this class should be used.

I ended up using some internal variables to reduce the amount of back and forth and amount of variables that needed to be passed on each function call.